### DISCUSSION TITLE
Is there a command for listing all the collections on a DevOps Server?

### DISCUSSION THREAD

**Discussion by Xergy** [DarqueWarrior/vsteam#512] [Q&A]
Is there a command for listing all the collections on a DevOps Server?

**Comment by SebastianSchuetze** [comment]
No there is not. It can only do what the API can do.

There is one cmdlet which can give you back the Collections / Organizations you have access to but not all perse. Also, because API is always done with the permissions of the user.

**Comment by sevaa** [comment]
The API can totally list all collections (that the current user can see): GET to `http://tfs.acme.com:8080/tfs/_apis/projectcollections?api-version=1.0-preview.2`

One problem is that this is a server level API (as opposed to project or collection) and this module doesn't readily support those in its shared REST client logic. The notion of Account, which the module treats as the ultimate TFS root, is a collection URL, not a server one, and the `_callAPI` function has no provision for working its way up from the account.

Matter of fact, I do have some ideas for supporting other server level methods (the extension gallery)...

**Reply by SebastianSchuetze** [reply]
Can you elaborate on an approach? Maybe on the level of the private functions?

Maybe we can come up with an idea to support it on the level when handling the login procedure (Set-VSTeamAccount).

**Reply by sevaa** [reply]
I don't have a ready made approach :)

The idea that Set-VSTeamAccount takes a collection URL (or an org; a root of all projects) is an extremely well established notion - were we to introduce an "OBTW this is a server URL" flag there, every single cmdlet would need to be amended to take an optional `Collection` parameter for those kinds of accounts, we probably don't want that. From my standpoint as a primarily on-prem user, it'd be a good idea - but only if the module was designed around that idea in the first place. :)

I'd rather introduce a "remove collection" flag for `_callAPI` that would slice the collection from the Account URL, and error out if that's impossible (i. e. if the current account is a legacy VSTS cloud URL with a trivial path, something like `https://myorg.visualstudio.com/`). Then have server level cmdlets use that.

Definitive guide on AzDevOps cloud URLs is here: https://learn.microsoft.com/en-us/azure/devops/extend/develop/work-with-urls?view=azure-devops&tabs=http On prem is a whole another story.
